<?xml version="1.0" ?>
<!--
    This launch file can be used to start the PaSLAM demonstration
    It starts the slam node, a rosbag and rviz
-->
<launch>


    <!-- to set the path from our customs models -->
    <env name="GAZEBO_MODEL_PATH" value="$(find pa_mobile_robot)/simulation/models/" />

    <!-- these are the arguments you can pass this launch file, for example paused:=true -->
    <arg name="paused"       default="false"/>
    <arg name="use_sim_time" default="true"/>
    <arg name="gui"          default="true"/>
    <arg name="headless"     default="false"/>
    <arg name="debug"        default="false"/>
    <arg name="node_start_delay"        default="10"/>
    <arg name="model"        default="$(find pa_mobile_robot)/simulation/urdf/differential_robot.xacro"/>



    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="world_name"   value="$(find pa_mobile_robot)/simulation/worlds/simple.world"/>
        <arg name="debug"        value="$(arg debug)" />
        <arg name="gui"          value="$(arg gui)" />
        <arg name="paused"       value="$(arg paused)"/>
        <arg name="use_sim_time" value="$(arg use_sim_time)"/>
        <arg name="headless"     value="$(arg headless)"/>
    </include>

    <param name="robot_description" command="$(find xacro)/xacro $(arg model)"/>

        <!-- SLAM -->
        <param name="max_cost_2add"       value="10000" /> 
        <param name="max_dist_2add"       value="0.5" />
        <param name="max_notadded"       value="5" />
        <param name="max_nochange_nandm"       value="2" />
        <param name="max_it_nandm"       value="20" />
        <param name="prob_map_pub_rate"       value="100" />
        <param name="cost_map_pub_rate"       value="0" />
        <param name="path_publishing_rate"       value="10" />
        <param name="publish_lidar_pose_tf"       value="true" />
        <param name="scan_topic_name"       value="/differiential_robot/scan" />
        <param name="path_topic_name"       value="/slam_path" />
        <param name="probmap_topic_name"       value="/paslam/probmap" />
        <param name="costmap_topic_name"       value="/paslam/costmap" />
        <param name="path_msg_queue"       value="100" />
        <param name="base_link_name"       value="paslam_baselink" />
        <param name="lidar_link_name"       value="laser" />
        <param name="map_link_name"       value="paslam_baselink" />
        <param name="resolution"       value="0.010" />
        <param name="height"       value="100" />
        <param name="width"       value="100" />
        <param name="cost_stamp_radius"       value="30" />
        <param name="max_belief"       value="100" />
        <param name="min_belief"       value="0" />
        <param name="add_belief"       value="25" />
        <param name="rem_belief"       value="10" />
        <param name="threshold_belief"       value="50" />
        <param name="initial_x"       value="0.0" />
        <param name="initial_y"       value="0.0" />
        <param name="initial_theta"       value="0.0" />
        <param name="lidar_msg_queue"       value="100" />
        
        
        <!-- EXPLORATION -->
        <param name="markers_frame_name"    value="paslam_baselink" />
        <param name="markers_namespace"     value="markers_namespace" />
        <param name="astar_path_topic_name"       value="/path_visualization_marker" />
        <param name="explo_map_topic_name"        value="/exploration_map" />
        <param name="explo_map_frame_name"        value="paslam_baselink" />
        <param name="twist_topic_name"      value="/differential_robot/cmd_vel" />
        <param name="linear_speed"          value="0.3" />
        <param name="angular_speed"         value="0.7" />
        <param name="publish_explomap_rate" value="20" />
        <param name="explo_map_resolution"        value="0.20" />
        <param name="update_twist_rate"     value="1" />


    <!-- NODES -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>

    <node name="joint_state_publisher" pkg="joint_state_publisher" type="joint_state_publisher"/>

    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model" respawn="false" output="screen"
          args="-z 0.0 -unpause -urdf -model TEST_Drive -param robot_description"/> 

    <node pkg="pa_mobile_robot" type="odom2path_node" name="true_odom_path_node" output="screen" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' ">
        <param name="base_link_name"       type="string" value="/differential_robot_odom"/>
        <param name="path_topic_name"      type="string" value="/odom2path/true_path"/>
        <param name="path_publishing_rate" type="int"    value="10"/>
        <param name="odom_topic_name"      type="string" value="/ground_truth/state"/>
    </node>

    <!-- The rviz node with the demo configuration -->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find pa_mobile_robot)/rviz/pa_exploration_slam_gazebo.rviz">
    </node>


    <!-- The SLAM node -->
    <node name="pa_exploration_slam_node" type= "pa_exploration_slam_node" pkg="pa_mobile_robot" output="screen" launch-prefix="bash -c 'sleep $(arg node_start_delay); $0 $@' ">
    </node>

</launch>
